### PR TITLE
Assessment processing

### PR DESCRIPTION
**Pull request by awisema**

### CODE REVIEW

File: db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class AddHmisFormProcessor < ActiveRecord::Migration[6.1]
+  def change
+    add_reference :hmis_assessment_details, :processor
+
+    create_table :hmis_assessment_processors do |t|
+      # HMIS keys are strings
```
Comment (eanders) [inline db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb]: Won't you need a data source id if you're storing the HMIS keys?

File: drivers/hmis/app/models/hmis/form/assessment_detail.rb [hunk 2]
```diff
@@ -9,6 +9,12 @@ class Hmis::Form::AssessmentDetail < ::GrdaWarehouseBase
   self.table_name = :hmis_assessment_details
   belongs_to :assessment, class_name: 'Hmis::Hud::Assessment'
   belongs_to :definition
+  belongs_to :processor, dependent: :destroy
+
+  def initialize(*)
+    super
+    build_processor
+  end
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/form/assessment_detail.rb]: Isn't this equivalent to:
```suggestion
  after_initialize :build_processor
```

File: drivers/hmis/app/models/hmis/form/processor.rb [hunk 3]
```diff
@@ -0,0 +1,196 @@
+###
+# Copyright 2016 - 2022 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+class Hmis::Form::Processor < ::GrdaWarehouseBase
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/form/processor.rb]: Should this be `AssessmentProcessor` since this isn't generic enough to handle services, events, and any other items, or is the idea that it eventually becomes generic?

File: drivers/hmis/app/models/hmis/form/processor.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  # assessment is accessed through the assessment_detail
+  belongs_to :health_and_dv, class_name: 'Hmis::Hud::HealthAndDv', primary_key: :HealthAndDVID, optional: true
+  belongs_to :income_benefit, class_name: 'Hmis::Hud::IncomeBenefit', primary_key: :IncomeBenefitID, optional: true
+  belongs_to :enrollment_coc, class_name: 'Hmis::Hud::EnrollmentCoc', primary_key: :EnrollmentCocID, optional: true
+  belongs_to :physical_disability, class_name: 'Hmis::Hud::Disability', primary_key: :DisabilityID, optional: true
+  belongs_to :developmental_disability, class_name: 'Hmis::Hud::Disability', primary_key: :DisabilitiesID, optional: true
+  belongs_to :chronic_health_condition, class_name: 'Hmis::Hud::Disability', primary_key: :DisabilitiesID, optional: true
+  belongs_to :hiv_aids, class_name: 'Hmis::Hud::Disability', primary_key: :DisabilitiesID, optional: true
+  belongs_to :mental_health_disorder, class_name: 'Hmis::Hud::Disability', primary_key: :DisabilitiesID, optional: true
+  belongs_to :substance_use_disorder, class_name: 'Hmis::Hud::Disability', primary_key: :DisabiliesID, optional: true
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/form/processor.rb]: These are going to need a `data_source_id` if we're joining on the HUD keys.  Could we just switch to using their internal IDs?

File: drivers/hmis/app/models/hmis/form/processor.rb [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+  # The items associated with the enrollments are all singletons, so return
+  # them if they already exist, otherwise create them
+  def enrollment_coc_factory
+    return enrollment_coc if enrollment_coc.present?
+
+    self.enrollment_coc = enrollment_factory.enrollment_cocs.
+      where(data_collection_stage: assessment_detail.data_collection_stage).
+      first_or_initialize(
+        enrollment_id: enrollment_factory.enrollment_id,
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/form/processor.rb]: Pretty sure all of these will need `data_source_id`

File: drivers/hmis/app/models/hmis/hud/processors/base.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  def initialize(processor)
+    @processor = processor
+  end
+
+  def hud_name(field)
+    field.underscore
+  end
+
+  def hud_type(field)
+    type = schema.fields[field].type
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/hud/processors/base.rb:27]: where is `schema` coming from?

File: drivers/hmis/app/models/hmis/hud/processors/enrollment_coc_processor.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module Hmis::Hud::Processors
+  class EnrollmentCocProcessor < Base
+    def process(field, value)
+      attribute_name = hud_name(field)
+      attribute_value = value
+      @processor.enrollment_coc_factory.assign_attributes(attribute_name => attribute_value)
+    end
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/hud/processors/enrollment_coc_processor.rb]: It feels like this could be dried up into the base class:
```suggestion
    def factory_method
      :enrollment_coc_factory
    end
    # move the next bit to the base class
    def process(field, value)
      attribute_name = hud_name(field)
      attribute_value = value
      @processor.send(factory_method).assign_attributes(attribute_name => attribute_value)
    end
```

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/processors/enrollment_coc_processor.rb]: Looking through more, this might be one of the exceptions, so maybe move the most common one up.

**Review by eanders — CHANGES_REQUESTED** [review]
Looks good! 
I'm concerned about the use of the HUD keys without an explicit data_source_id, otherwise, just some possible optimizations.

File: db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      t.string :enrollment_coc_id
+      t.string :health_and_dv_id
+      t.string :income_benefit_id
+      t.string :physical_disability_id
+      t.string :developmental_disability_id
+      t.string :chronic_health_condition_id
+      t.string :hiv_aids_id
+      t.string :mental_health_disorder_id
+      t.string :substance_use_disorder_id
```
Comment (gigxz) [inline db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb]: We will need keys for YouthEducationStatus and EmploymentEduction records here eventually, if you are reworking the migrations anyway and want to add them now

File: drivers/hmis/app/graphql/mutations/submit_assessment.rb [hunk 9]
```diff
@@ -52,6 +52,7 @@ def resolve(assessment_id: nil, enrollment_id: nil, form_definition_id: nil, val
 
       if assessment.valid? && assessment.assessment_detail.valid?
         assessment.assessment_detail.save!
+        assessment.assessment_detail.processor.run!
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/submit_assessment.rb]: We'll need to get validation errors for each record out of this

File: drivers/hmis/app/models/hmis/form/processor.rb (hunk 4 above)
Comment (awisema) [reply drivers/hmis/app/models/hmis/form/processor.rb]: Good question -- I forgot we had Rails IDs...

File: drivers/hmis/app/models/hmis/form/processor.rb [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+  def health_and_dv_factory
+    return health_and_dv if health_and_dv.present?
+
+    self.health_and_dv = enrollment_factory.health_and_dvs.
+      where(data_collection_stage: assessment_detail.data_collection_stage).
+      first_or_initialize(
+        enrollment_id: enrollment_factory.enrollment_id,
+        personal_id: enrollment_factory.client.personal_id,
+        information_date: assessment_detail.assessment.assessment_date,
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/form/processor.rb]: The information date will be coming from hudValues. Either as a single `informationDate` key (preferable to reduce form definition complexity) or as multiple fields like `IncomeBenefit.informationDate` etc. 

I think we should probably store the InformationDate directly on the AssessmentDetail too, so that it can be used as a column in the assessments table for example.

File: drivers/hmis/app/models/hmis/form/processor.rb [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        user_id: assessment_detail.assessment.user_id,
+      )
+  end
+
+  def health_and_dv_factory
+    return health_and_dv if health_and_dv.present?
+
+    self.health_and_dv = enrollment_factory.health_and_dvs.
+      where(data_collection_stage: assessment_detail.data_collection_stage).
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/form/processor.rb]: I think the Information Date needs to be included here because you can have multiple records with the same Data Collection Stage (Update and Annual)

File: drivers/hmis/app/models/hmis/hud/processors/base.rb (hunk 6 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/processors/base.rb:27]: it's the GraphQL type, defined on each of the subclasses of `Hmis::Hud::Processors::Base`

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/processors/base.rb:27]: Thanks.  Should this be defined as an abstract class then?

File: drivers/hmis/spec/models/hmis/form/processor_spec.rb [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    assessment.assessment_detail.processor.run!
+    assessment.save_not_in_progress
+
+    expect(assessment.enrollment.income_benefits.count).to eq(1)
+
+    income_benefits = assessment.enrollment.income_benefits.first
+    expect(income_benefits.income_from_any_source).to eq(1)
+    expect(income_benefits.earned).to eq(1)
```
Comment (gigxz) [inline drivers/hmis/spec/models/hmis/form/processor_spec.rb]: This is correct behavior (`true` saved as `1`) but how is it working? I don't understand because the GraphQL Schema type is Boolean and I'm not seeing a transformation.

Can you add test cases to cover that `false` gets saved as `0` and `nil` gets saved as `99`?

File: drivers/hmis/spec/models/hmis/form/processor_spec.rb [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+      'DisabilityGroup.physicalDisability' => 'YES',
+      'DisabilityGroup.physicalDisabilityIndefiniteAndImpairs' => 'YES',
+      'DisabilityGroup.developmentalDisability' => 'NO',
+      'DisabilityGroup.chronicHealthCondition' => 'YES',
+      'DisabilityGroup.chronicHealthConditionIndefiniteAndImpairs' => 'NO',
+      'DisabilityGroup.hivAids' => 'YES',
+      'DisabilityGroup.mentalHealthDisorder' => 'NO',
+      'DisabilityGroup.substanceUseDisorder' => 'BOTH_ALCOHOL_AND_DRUG_USE_DISORDERS',
+      'DisabilityGroup.substanceUseDisorderIndefiniteAndImpairs' => 'YES',
+      'DisabilityGroup.disablingCondition' => 'YES',
```
Comment (gigxz) [inline drivers/hmis/spec/models/hmis/form/processor_spec.rb]: I think we still need to determine whether 99s should be filled in by the frontend before sending back, or if they should be added by the processor.

For example say someone enters "No" for physical disability, "Yes" for hiv/aids, and leaves everything else blank. It (currently) results in the following
```
  "DisabilityGroup.physicalDisability": "NO",
  "DisabilityGroup.hivAids": "YES",
  "DisabilityGroup.disablingCondition": "YES"
```

* The physical disability record's "IndefiniteAndImpairs" field should be saved as `99`, not `nil`
* Should we still create all 6 disability records (filled with 99s) or should we only created the 2 that were specified?

Definitely for a follow-up discussion/PR, no changes requested on that here.

File: drivers/hmis/spec/models/hmis/form/processor_spec.rb [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    expect(disabilities.find_by(disability_type: 5).indefinite_and_impairs).to eq(1)
+    # Developmental Disability
+    expect(disabilities.find_by(disability_type: 6).disability_response).to eq(0)
+    expect(disabilities.find_by(disability_type: 6).indefinite_and_impairs).to eq(nil)
+    # Substance Use
+    expect(disabilities.find_by(disability_type: 10).disability_response).to eq(3)
+  end
+
+  describe 'updating existing assessment' do
+    it "doesn't touch an existing value, if it isn't listed (but applies the listed fields)" do
```
Comment (gigxz) [inline drivers/hmis/spec/models/hmis/form/processor_spec.rb]: I think you're right that it should behave this way, but I will need to make changes to the frontend to make it work.


For example if I first enter earned amount 50 and submit:
```
"IncomeBenefit.incomeFromAnySource": "YES",
"IncomeBenefit.earned": true,
"IncomeBenefit.earnedAmount": 50
```
and then go back and check "NO" for any income and re-submit, it will send:
```
"IncomeBenefit.incomeFromAnySource": "NO",
```

So I need to update frontend to send explicit nulls for fields that become disabled.

File: drivers/hmis/app/models/hmis/hud/processors/base.rb (hunk 6 above)
Comment (awisema) [reply drivers/hmis/app/models/hmis/hud/processors/base.rb:27]: Probably -- but these are utility classes, `self.abstract_class = true` is a AR construct. I'll put in
```
def schema
  raise 'Implement in sub-class'
end
```

**Review by gigxz — APPROVED** [review]
This looks great! Thank you! Most of my comments are open-ended questions that can be addressed after merging this in.
I think most pressing is probably to implement information date correctly, but happy to do that in a second iteration.

File: drivers/hmis/spec/models/hmis/form/processor_spec.rb (hunk 13 above)
Comment (eanders) [reply drivers/hmis/spec/models/hmis/form/processor_spec.rb]: We should definitely create all 6 records with 99s or nils.  I think not having some of them may really impact reporting down the road.

File: drivers/hmis/app/models/hmis/form/processor.rb (hunk 5 above)
Comment (awisema) [reply drivers/hmis/app/models/hmis/form/processor.rb]: Turns out the `enrollment_id` and `data_source_id` get set via Rails magic (they are declared as the keys in the `has_many`) -- so I went the other way, and took `enrollment_id` out.

File: db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb (hunk 8 above)
Comment (awisema) [reply db/warehouse/migrate/20230112142317_add_hmis_form_processor.rb]: I'd rather do that as a separate migration when we add the appropriate processors as well. That OK with you?

File: drivers/hmis/spec/models/hmis/form/processor_spec.rb (hunk 12 above)
Comment (awisema) [reply drivers/hmis/spec/models/hmis/form/processor_spec.rb]: These rely on the AR database translations, which seem to work for the cases we have, so I didn't want to complicate the code/

There is `false` in this test (`'IncomeBenefit.otherInsurance' => false`), and I think we decided the `nil` is clearing the field, so you will need to send a `99` if you want to store one. Is this incorrect?

Comment (gigxz) [reply drivers/hmis/spec/models/hmis/form/processor_spec.rb]: Hmm right. I didn't realize it was a nullable 1.7 field.. I wonder why HUD does that. I will think about this for all the 1.7 fields

File: drivers/hmis/app/models/hmis/form/processor.rb (hunk 11 above)
Comment (awisema) [reply drivers/hmis/app/models/hmis/form/processor.rb]: After discussion, we decided we don't need the `where`, and replaced this with a `build`.

File: drivers/hmis/app/models/hmis/form/processor.rb (hunk 10 above)
Comment (awisema) [reply drivers/hmis/app/models/hmis/form/processor.rb]: After discussion, we decided this will come from the assessment as implemented, and additionally, we need to update the `information_date` to match the `assessment_date` whenever it changes.

**Review by gigxz — APPROVED** [review]

File: db/structure.sql [hunk 15]
```diff
@@ -276,6 +276,7 @@ CREATE TABLE public.activity_logs (
 --
 
 CREATE SEQUENCE public.activity_logs_id_seq
+    AS integer
```
Comment (eanders) [inline db/structure.sql]: What's causing this file to change?  Anything intentional?

Comment (awisema) [reply db/structure.sql]: No... I'll revert it.

**Review by eanders — APPROVED** [review]
